utils/minnesota.py

```diff
@@ -0,0 +1,423 @@
+        """
+
+        df1 = df.copy(deep=True)
+        df1 = df1[MN_NONCANDIDATE_CONTRIBUTION_COL]
+        df1 = df1.rename(columns=MN_NONCANDIDATE_CONTRIBUTION_MAP)
+        df1["recipient_first_name"] = None
+        df1["recipient_last_name"] = None
+        df1["donor_type"] = df1["donor_type"].str.upper()
+        df1["donor_first_name"] = None
+        df1["donor_last_name"] = None
+        df1["state"] = "MN"
+        df1["office_sought"] = None
+        df1["amount"] = pd.to_numeric(df1["amount"], errors="coerce")
+        df1["donor_id"] = df1["donor_id"].fillna(0).astype("int64")
+        df1["inkind_amount"] = pd.to_numeric(df1["inkind_amount"], errors="coerce")
+        df1["transaction_type"] = np.where(
+            (df1["inkind_amount"].notna()) & (df1["inkind_amount"] != 0),
+            "in-kind",
+            None,
+        )
+
+        return df1
+
+    def preprocess_expenditure(self, df: pd.DataFrame) -> pd.DataFrame:
+        """
+        Helper function to preprocess MN expenditure dataset
+
+        Args:
+            df (DataFrame): MN expenditure DataFrames
+        Returns:
+            DataFrame: Preprocessed MN expenditure DataFrames
+        """
+
+        df1 = df.copy(deep=True)
+        df1 = df1[MN_INDEPENDENT_EXPENDITURE_COL]
+        df1 = df1.rename(columns=MN_INDEPENDENT_EXPENDITURE_MAP)
+        # Donors and recipients are both organization, only have full name
+        df1["recipient_first_name"] = None
+        df1["recipient_last_name"] = None
+        df1["donor_first_name"] = None
+        df1["donor_last_name"] = None
+        df1["recipient_id"] = df1["donor_id"].fillna(0).astype("int64")
+        # Negate the contribution amount if it's against the recipient
+        df1.loc[df1["For /Against"] == "Against", "amount"] = -df1["amount"]
+        df1["inkind_amount"] = None
+        df1 = df1.drop(columns=["For /Against"])
+        df1["recipient_type"] = "PCF"  # recipient: affected political committee
+        df1["office_sought"] = None
+
+        return df1
+
+    def preprocess(self, filepaths_list: list[str]) -> list[pd.DataFrame]:
+        """
+        Preprocesses MN data and returns a list of a single dataframe combining
+        both contributions and expenditures
+
+        Inputs:
+            filepaths_list (list): list of absolute filepaths of contribution
+            and expenditure csv files
+
+        Order: 10 candidate-recipient contribution files (AG, AP, DC, GC, House,
+            SA, SC, Senate, SS, ST), 1 noncandidate-recipient contribution file,
+            and 1 expenditure file
+
+        Naming Convention: root + data folder (+ candidate folder) + file name
+
+        Returns:
+            A list containing 1 preprocessed DataFrame
+        """
+
+        candidate_dfs = []
+        processed_dfs = []
+        # candidate-recipient contribution data
+        for filepath1 in filepaths_list[:-2]:
```

Comment:
you can read the file in, preprocess it, and then append it and get rid of one of the for-loops